### ISSUE TITLE
Base64.EncodeToUtf8 behavior changed between netcore2.1 and netcore3.0

### ISSUE THREAD

**Issue by coenm** [dotnet/corefx#42236]
The following call

```csharp
var status = System.Buffers.Text.Base64.EncodeToUtf8(Span<byte>.Empty, new byte[42], out _, out _, isFinalBlock:false);
```

returns `OperationStatus.NeedMoreData` running netcore2.1 and `OperationStatus.Done` in netcore3.0

Is this intentionally? Is this a bug? Should this be documented?

Example can be found at:

```bash
git clone https://github.com/coenm/Base64Issue.git

dotnet test Base64Difference/Base64Difference.csproj
```
The test will pass in netcore2.1 and fail in netcore3.0

**Comment by gfoidl** [comment]
With https://github.com/dotnet/corefx/blob/fd454581475dd32b88ed45d38976855f24bd6751/src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs#L29 I would say to is a bug, as "more data will follow".

I introduced this in https://github.com/dotnet/corefx/pull/34529, so I'll prepare a fix (+ test, as there is none, obviously).

**Comment by gfoidl** [comment]
Same applies to `Base64.DecodeFromUtf8`.

**Comment by GrabYourPitchforks** [comment]
I believe the 3.0 behavior to be correct. `OperationStatus.Done` means "this method has consumed the entirety of the source data without error, and there was enough space in the destination buffer to hold the resulting output." For an empty source buffer, this is trivially true.

**Comment by gfoidl** [comment]
@GrabYourPitchforks this was my initial thought also, but then I looked at `isFinalBlock: false` and that changed my mind.

**Comment by GrabYourPitchforks** [comment]
The documentation for the _isFinalBlock_ parameter is incorrect. The documentation should instead read "Set to _true_ when the source buffer contains the entirety of the data to encode. Set to _false_ if this method is being called in a loop and if more input data may follow. At the end of the loop, call this (potentially with an empty source buffer) passing _true_.."

**Comment by gfoidl** [comment]
Yep, this seems fine and natural to me.

`isFinalBlock: false` with empty input is not very natural, and I can't imagine a proper use-case for this.
But if any user has such a scenario the .NET Core 2.1 behavior would be better.
So "termination" (i.e. `Done`) can be achieved with `isFinalBlock: true` only.

I don't have a strong opinion here either.

**Comment by GrabYourPitchforks** [comment]
> So "termination" (i.e. `Done`) can be achieved with `isFinalBlock: true` only.

This is incorrect. `OperationStatus` return values do not reflect the status of the entire operation, which might consist of multiple calls to the `OperationStatus`-returning step method. `OperationStatus` return values only reflect the status of one particular iteration: the immediate method call which returned the value in question. And this particular iteration is `Done` because the entirety of the source buffer was consumed and there was enough space in the destination buffer to hold the resulting value without error.

The behavior I describe in the above paragraph is the intended behavior of `OperationStatus` and reflects the behavior of other `OperationStatus`-based APIs we expose. For example:

```cs
using System;
using System.Text.Unicode;

namespace App
{
    class Program
    {
        static void Main(string[] args)
        {
            byte[] output = new byte[32];
            var status = Utf8.FromUtf16(string.Empty, output, out int charsread, out int bytesWritten, isFinalBlock: false);
            Console.WriteLine(status); // outputs "Done"
        }
    }
}
```

**Comment by gfoidl** [comment]
> And this particular iteration is `Done` because the entirety of the source buffer was consumed and there was enough space in the destination buffer to hold the resulting value without error.

Then the whole `isFinalBlock: false`-story in `System.Buffers.Text.Base64` is broken.
```c#
using System;
using System.Buffers.Text;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            Span<byte> input = stackalloc byte[3] { 0x1, 0x2, 0x3 };
            Span<byte> output = stackalloc byte[4];

            var status = Base64.EncodeToUtf8(input, output, out int read, out int written, isFinalBlock: false);
            Console.WriteLine(status);  // outputs "NeedMoreData"
        }
    }
}
```

To follow your statement of `Done`, the result should be changed to cover these cases for `isFinalBlock: false`:

| Case | `OperationStatus` | Comment |
| -- | -- | -- |
| input % 3 == 0, output large enough | `Done` | |
| input % 3 != 0, output large enough | `NeedMoreData` | |
| output not large enough | `DestinationTooSmall` | encoded as much as possible |

Or has `OperationStatus` a special meaning in base64 with regards of padding etc? No, that would not fit to your statement (note to myself), and padding comes only for `isFinalBlock: true` into play.

**Comment by GrabYourPitchforks** [comment]
Yes, that is a bug. In your scenario the result should be `OperationStatus.Done`.

**Comment by gfoidl** [comment]
Just checked the [initial API proposal](https://github.com/dotnet/corefx/issues/24568#issuecomment-338833370), and that is in line with your view:

> * Done - the entire input buffer has been consumed, i.e. `consumed == input.Length`, the operation is complete.
> * NeedMoreData - the input is partially consumed, up to the last valid chunk of the input that could be consumed

--- 

> Yes, that is a bug

Good to know. This bug is then around since the base64-apis got introduced...
I'm going to address this (in the next couple of days).

**Comment by GrabYourPitchforks** [comment]
As another example of why `OperationStatus.Done` is the correct return value, consider an API that would perform hex encoding:

```cs
public static OperationStatus HexEncode(ReadOnlySpan<byte> source, Span<char> destination, out int bytesConsumed, out int charsWritten);
```

This API doesn't take an `isFinalBlock` parameter at all because it's unnecessary. One byte of input always translates to two chars of output, and there's no special processing which needs to take place for the "final" chunk of data. This API can be used in a one-shot manner, or it can be used in a loop to process chunks of a very large buffer.

In such an API, when would we ever know to return `OperationStatus.NeedsMoreData`? Such a return value is nonsensical in such an API. Either the input was fully consumed (`Done`), or it wasn't (`DestinationTooSmall`).

The reverse (`HexDecode`) operation _would_ need an `isFinalBlock` API because two input chars go to a single byte, and we'd need to distinguish "you need to provide more input data" (`NeedsMoreData`) vs. "you gave me an odd-length input and told me this was the end of the stream" (`InvalidData`).

**Comment by gfoidl** [comment]
Thanks, I got it. 

Had a mental connection of `NeedsMoreData` with `isFinalBlock: false`. Now in hindsight this doesn't make sense at all, because as caller when I know / set `isFinalBlock: false` then I don't need a return value telling me that more data is expected :wink: (that's why `false` was used), so `Done` is what I expect, to know that my provided data was encoded successfully. Then more data will be sent, until `isFinalBlock: true` will be set with the remaining (potential empty) data.

**Comment by GrabYourPitchforks** [comment]
Thanks for finding the issue with the non-empty buffers! I opened a tracking item for it.

**Comment by GrabYourPitchforks** [comment]
Relabeling this with _documentation_ and _breaking-change_ so that we know we should doc it.

**Comment by ahsonkhan** [comment]
I am not certain if we want to take this breaking change. There are two questions here:
1) Is the .NET Core 2.1 behavior intended and correct? 
2) Let's assume it isn't, should we make the breaking change or revert back to the original behavior?

Regarding 1, I think the `isFinalBlock=false` changes the circumstance. I am not 100% sold yet that if someone passes false there, we should be returning Done, so to me, NeedsMoreData makes more sense. Done should only be returned when the user sets `isFinalBlock=true` (from OpStatus docs for Done ["... and the operation is complete"](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.operationstatus?view=netcore-3.0)). It might be helpful to write/share some calling code where the partial span is being passed-in, in a loop, and what the user might expect from the return values between multiple iterations (see below). I would have considered `OperationStatus = Done` as a "break out of the loop" condition. In that case, someone passing in false for isFinalBlock as there "last iteration" could be an indication of a bug in their code. The user should only break out of their loop after the last call had `isFinalBlock=true` passed in to it. That way, the break condition can be based on `OperationStatus == Done`.

Let's settle on 1 first, but assuming we think Done is the correct behavior, should we be changing the semantics for this particular API and take the breaking change, or leave the behavior as it was in .NET Core 2.1? I can see some value in us being consistent with the semantics of other OpStatus-based APIs (particularly newer ones), but I don't know if the risk to potentially break people who rely on the current behavior is worth the benefit. It might be OK for keep the semantics of this particular API as it originally shipped. Currently, 3.0 has different behavior for empty buffer than 2.1, but same behavior for non-empty buffers. 
As an example: consider code like this Debug.Assert breaking:
https://github.com/dotnet/corefx/blob/b456415bca6317b1ad6f98c6475ca59a20c57101/src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs#L48

```C#
static void Main(string[] args)
{
    byte[] input = new byte[1_000];
    byte[] destination = new byte[2_000];
    var random = new Random(42);
    random.NextBytes(input);

    var source = new MemoryStream(input);
    
    byte[] buffer = new byte[128];

    int written = 0;
    int leftover = 0;

    OperationStatus status = OperationStatus.NeedMoreData;

    while (status != OperationStatus.Done)
    {
        int bytesRead = source.Read(buffer.AsSpan(leftover));
        bool isFinalBlock = bytesRead == 0;

        int bytesToProcess = bytesRead + leftover;

        status = Base64.EncodeToUtf8(
            buffer.AsSpan(0, bytesToProcess), 
            destination.AsSpan(written), 
            out int bytesConsumed, 
            out int bytesWritten, 
            isFinalBlock);

        written += bytesWritten;
        leftover = bytesToProcess - bytesConsumed;

        // For this case, otherwise, re-grow destination or  throw for invalid/etc.
        Debug.Assert(status != OperationStatus.DestinationTooSmall &&
                    status != OperationStatus.InvalidData);

        if (status == OperationStatus.Done)
        {
            Debug.Assert(leftover == 0);
        }

        if (status == OperationStatus.NeedMoreData)
        {
            Debug.Assert(leftover != 0 || isFinalBlock == false);
        }
    }

    Console.WriteLine(Encoding.UTF8.GetString(destination.AsSpan(0, written)));
}
```

**Comment by GrabYourPitchforks** [comment]
> Is the .NET Core 2.1 behavior intended and correct?

I think this thread has presented overwhelming evidence that the 3.0 behavior is correct for empty input buffers. This is backed up by the behavior of other `OperationStatus`-based APIs, a thought experiment involving future `OperationStatus` APIs that don't accept an _isFinalBlock_ parameter, and references back to the original issue description which clearly indicate that the return value was intended to be a per-iteration indicator instead of a cross-iteration indicator. The 3.0 behavior for empty input buffers also follows the pattern established by the closest thing .NET has had to the `OperationStatus` concept: the _out bool completed_ parameter for APIs like `Encoder/Decoder.Convert`.

Your argument that we should preserve the behavior for compat purposes is compelling. I'd argue that we shouldn't change this behavior in a point-release, but in a major release there's certainly benefit to making the API behaviors consistent across all `OperationStatus`-returniing APIs.

To the code sample, for something like Base64-encoding transforms callers generally shouldn't be looking at the returned `OperationStatus` anyway if an infinite sink is available. (The pattern below is generally applicable to any API that is contracted never to return `OperationStatus.InvalidData` and where the caller already knows that the scratch buffer is big enough to make some sort of forward progress on every call.)

```cs
public static void CopyToAndBase64Encode(Stream inputStream, IBufferWriter<byte> output)
{
    Span<byte> scratchBuffer = new byte[2048];
    int leftoverDataLength = 0;

    while (true) {

        // Read data from the input stream

        int bytesReadFromStream = inputStream.Read(scratchBuffer.Slice(leftoverDataLength));
        if (bytesReadFromStream == 0)
        {
            break; // EOF, process leftover data
        }

        // Encode into the IBufferWriter

        Span<byte> destination = output.GetSpan();
        Base64.EncodeToUtf8(
            scratchBuffer.Slice(leftoverDataLength + bytesReadFromStream),
            destination,
            out int bytesConsumedByEncode,
            out int bytesWrittenByEncode,
            isFinalBlock: false);
        output.Advance(bytesWrittenByEncode);

        // Move unprocessed data back to the beginning of the scratch buffer

        Span<byte> unprocessedData = scratchBuffer[bytesConsumedByEncode..(leftoverDataLength + bytesReadFromStream)];
        unprocessedData.CopyTo(scratchBuffer);
        leftoverDataLength = unprocessedData.Length;
    }

    // Final transform

    Span<byte> finalDestination = output.GetSpan(Base64.GetMaxEncodedToUtf8Length(leftoverDataLength));
    OperationStatus status = Base64.EncodeToUtf8(
        scratchBuffer.Slice(0, leftoverDataLength),
        finalDestination,
        out int bytesReadByFinalEncode,
        out int bytesWrittenByFinalEncode,
        isFinalBlock: true);
    Debug.Assert(status == OperationStatus.Done);
    Debug.Assert(bytesReadByFinalEncode == leftoverDataLength);
    output.Advance(bytesWrittenByFinalEncode);
}
```

**Comment by GrabYourPitchforks** [comment]
For more info on why the calling code in my sample is written the way it is (without even looking at `OperationStatus`), check out the discussion at https://youtu.be/BEBq3__WfDc?t=6134 (timestamp 1:42:14).

**Comment by GrabYourPitchforks** [comment]
Also, for further context check out recording at https://youtu.be/BEBq3__WfDc?t=5160 (timestamp 1:26:00), where we discuss the interplay between `isFinalBlock` and `OperationStatus.NeedsMoreData`. The discussion there revolves around that it's intended to resolve an ambiguity regarding "I don't know whether I need to give this data special treatment because I don't know if I'm going to be called again." The cases we're describing here do not have such an ambiguity, as empty input always corresponds to empty output regardless of whether it's in the middle or at the end of the buffer.

**Comment by bartonjs** [comment]
The API docs for OperationStatus (https://source.dot.net/#System.Private.CoreLib/shared/System/Buffers/OperationStatus.cs,52cc599542def2d4) suggest that Done is the right answer, too.

* `Done`: The entire input buffer has been processed and the operation is complete.
  * Sounds like if `bytesConsumed` == `source.Length` that Done is the correct answer.
  * Yeah, there's a question of what an "operation" means, but in the context of a stateless method (state is in the caller) "this call" seems natural.
* `DestinationTooSmall`: The input is partially processed, up to what could fit into the destination buffer.
  * "partially" => `bytesConsumed` < `source.Length`
* `NeedMoreData`: The input is partially processed, up to the last valid chunk of the input that could be consumed.
  * "partially", again.
* `InvalidData`: The input contained invalid bytes which could not be processed.
  * Not really relevant to this case, but I'm a completionist.

OperationStatus is really just a Try- of Try-s.  For `EncodeToUtf8(..., isFinalBlock: false)` the flow, to me, is:

* If (a mythical) `byte[] EncodeToUtf8ForbidPadding(byte[] data)` would throw (other than OOM): `NeedMoreData`.
* If (a mythical) `TryEncodeToUtf8ForbidPadding(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten)` would return false: `DestinationTooSmall`.
* Else return `Done`.

For `isFinalBlock: true`, call a different method pair, `Try`/`EncodeToUtf8WithPadding`.  The array-returning one will never throw, so `NeedMoreData` is eliminated from the return repertoire.

**Comment by GrabYourPitchforks** [comment]
> Sounds like if `bytesConsumed == source.Length` that Done is the correct answer.

There is potentially one exception to this: an API could return `DestinationTooSmall` even if it has consumed the entire input buffer. An example of such an API would be something that writes a marker to the end of the destination stream saying "I'm finished!", such as might occur within a CBC-style cipher.

```cs
Span<byte> input = new byte[16];
Span<byte> output = new byte[16];

// Presumably returns 'DestinationTooSmall' because even though it did consume the entire input buffer,
// it needs additional room to write the 16 padding bytes [ 0x10 ].

OperationStatus status = EncipherCbc(input, output, out int bytesConsumed, out int bytesWritten, isFinalBlock: true);
Debug.Assert(status == OperationStatus.DestinationTooSmall);
Debug.Assert(bytesConsumed == 16);
Debug.Assert(bytesWritten == 16);

input = input.Slice(bytesConsumed); // input is now empty
output = GetAdditionalOutputBuffer();

status = EncipherCbc(input /* empty */, output, out bytesConsumed, out bytesWritten, isFinalBlock: true);
Debug.Assert(status == OperationStatus.Done);
Debug.Assert(bytesConsumed == 0);
Debug.Assert(bytesWritten == 16); // final padding written
```

**Comment by bartonjs** [comment]
Fair enough. If `U[] DoTheThing(Array<T>.Empty)` produces a non-empty thing, then "consumed == source.Length" might not be `Done` :smile:
